packages/kbn-url-state/index.ts

```diff
@@ -6,4 +6,109 @@
  * Side Public License, v 1.
  */
 
-export { useSyncToUrl } from './use_sync_to_url';
+import { useCallback, useEffect, useState } from 'react';
+import { encode, decode, RisonValue } from '@kbn/rison';
+import { stringify, parse } from 'query-string';
+
+interface StateCache<T> {
+  namespaces: Record<string, Record<string, T | undefined>>;
+  timeoutHandle: number;
+}
+
+/**
+ * Temporary cache for state stored in the URL. This will be serialized to the URL
+ * in a single batched update to avoid excessive history entries.
+ */
+const cache: StateCache<unknown> = {
+  namespaces: {},
+  timeoutHandle: 0,
+};
+
+const CUSTOM_URL_EVENT = 'url:update' as const;
+
+// This is a list of events that can trigger a render.
+const URL_CHANGE_EVENTS: string[] = ['popstate', CUSTOM_URL_EVENT];
+
+/**
+ * This hook stores state in the URL, but with a namespace to avoid collisions with other values in the URL.
+ * It also batches updates to the URL to avoid excessive history entries.
+ * With it, you can store state in the URL and have it persist across page refreshes.
+ * The state is stored in the URL as a Rison encoded object.
+ *
+ * Example: when called like this `const [value, setValue] = useUrlState<boolean>('myNamespace', 'myKey');`
+ * the state will be stored in the URL like this: `?myNamespace=(myKey:!n)`
+ *
+ * State is not cleared from the URL when the hook is unmounted and this is by design.
+ * If you want it to be cleared, you can do it manually by calling `setValue(undefined)`.
+ *
+ * @param urlNamespace actual top level query param key
+ * @param key sub key of the query param
+ */
+export const useUrlState = <T = unknown>(urlNamespace: string, key: string) => {
```

Comment:
@lgestc , love the simplicity of the API. 

I understand that you are delaying updates to the next `microtask` with `setTimeout` but I don't 100% understand, how does it batches the state updates.

Also, I think something like `debounce` can provide a clearer api for batching very quick successive updates,

Reply:
so the changes are first appended to some global object, and then flushed to the url, but we are going to drop this custom thing in favor of already made url storage solution linked somewhere here